// windmill-client/dist/core/request.mjs
import { ApiError } from "./ApiError.mjs";
import { CancelablePromise } from "./CancelablePromise.mjs";

//#region src/core/request.ts
const isString = (value) => {
	return typeof value === "string";
};
const isStringWithValue = (value) => {
	return isString(value) && value !== "";
};
const isBlob = (value) => {
	return value instanceof Blob;
};
const isFormData = (value) => {
	return value instanceof FormData;
};
const base64 = (str) => {
	try {
		return btoa(str);
	} catch (err) {
		return Buffer.from(str).toString("base64");
	}
};
const getQueryString = (params) => {
	const qs = [];
	const append = (key, value) => {
		qs.push(`${encodeURIComponent(key)}=${encodeURIComponent(String(value))}`);
	};
	const encodePair = (key, value) => {
		if (value === void 0 || value === null) return;
		if (value instanceof Date) append(key, value.toISOString());
		else if (Array.isArray(value)) value.forEach((v) => encodePair(key, v));
		else if (typeof value === "object") Object.entries(value).forEach(([k, v]) => encodePair(`${key}[${k}]`, v));
		else append(key, value);
	};
	Object.entries(params).forEach(([key, value]) => encodePair(key, value));
	return qs.length ? `?${qs.join("&")}` : "";
};
const getUrl = (config, options) => {
	const encoder = config.ENCODE_PATH || encodeURI;
	const path = options.url.replace("{api-version}", config.VERSION).replace(/{(.*?)}/g, (substring, group) => {
		if (options.path?.hasOwnProperty(group)) return encoder(String(options.path[group]));
		return substring;
	});
	const url = config.BASE + path;
	return options.query ? url + getQueryString(options.query) : url;
};
const getFormData = (options) => {
	if (options.formData) {
		const formData = new FormData();
		const process = (key, value) => {
			if (isString(value) || isBlob(value)) formData.append(key, value);
			else formData.append(key, JSON.stringify(value));
		};
		Object.entries(options.formData).filter(([, value]) => value !== void 0 && value !== null).forEach(([key, value]) => {
			if (Array.isArray(value)) value.forEach((v) => process(key, v));
			else process(key, value);
		});
		return formData;
	}
	return void 0;
};
const resolve = async (options, resolver) => {
	if (typeof resolver === "function") return resolver(options);
	return resolver;
};
const getHeaders = async (config, options) => {
	const [token, username, password, additionalHeaders] = await Promise.all([
		resolve(options, config.TOKEN),
		resolve(options, config.USERNAME),
		resolve(options, config.PASSWORD),
		resolve(options, config.HEADERS)
	]);
	const headers = Object.entries({
		Accept: "application/json",
		...additionalHeaders,
		...options.headers
	}).filter(([, value]) => value !== void 0 && value !== null).reduce((headers$1, [key, value]) => ({
		...headers$1,
		[key]: String(value)
	}), {});
	if (isStringWithValue(token)) headers["Authorization"] = `Bearer ${token}`;
	if (isStringWithValue(username) && isStringWithValue(password)) {
		const credentials = base64(`${username}:${password}`);
		headers["Authorization"] = `Basic ${credentials}`;
	}
	if (options.body !== void 0) {
		if (options.mediaType) headers["Content-Type"] = options.mediaType;
		else if (isBlob(options.body)) headers["Content-Type"] = options.body.type || "application/octet-stream";
		else if (isString(options.body)) headers["Content-Type"] = "text/plain";
		else if (!isFormData(options.body)) headers["Content-Type"] = "application/json";
	}
	return new Headers(headers);
};
const getRequestBody = (options) => {
	if (options.body !== void 0) if (options.mediaType?.includes("application/json") || options.mediaType?.includes("+json")) return JSON.stringify(options.body);
	else if (isString(options.body) || isBlob(options.body) || isFormData(options.body)) return options.body;
	else return JSON.stringify(options.body);
	return void 0;
};
const sendRequest = async (config, options, url, body, formData, headers, onCancel) => {
	const controller = new AbortController();
	let request$1 = {
		headers,
		body: body ?? formData,
		method: options.method,
		signal: controller.signal
	};
	if (config.WITH_CREDENTIALS) request$1.credentials = config.CREDENTIALS;
	for (const fn of config.interceptors.request._fns) request$1 = await fn(request$1);
	onCancel(() => controller.abort());
	return await fetch(url, request$1);
};
const getResponseHeader = (response, responseHeader) => {
	if (responseHeader) {
		const content = response.headers.get(responseHeader);
		if (isString(content)) return content;
	}
	return void 0;
};
const getResponseBody = async (response) => {
	if (response.status !== 204) try {
		const contentType = response.headers.get("Content-Type");
		if (contentType) {
			const binaryTypes = [
				"application/octet-stream",
				"application/pdf",
				"application/zip",
				"audio/",
				"image/",
				"video/"
			];
			if (contentType.includes("application/json") || contentType.includes("+json")) return await response.json();
			else if (binaryTypes.some((type) => contentType.includes(type))) return await response.blob();
			else if (contentType.includes("multipart/form-data")) return await response.formData();
			else if (contentType.includes("text/")) return await response.text();
		}
	} catch (error) {
		console.error(error);
	}
	return void 0;
};
const catchErrorCodes = (options, result) => {
	const errors = {
		400: "Bad Request",
		401: "Unauthorized",
		402: "Payment Required",
		403: "Forbidden",
		404: "Not Found",
		405: "Method Not Allowed",
		406: "Not Acceptable",
		407: "Proxy Authentication Required",
		408: "Request Timeout",
		409: "Conflict",
		410: "Gone",
		411: "Length Required",
		412: "Precondition Failed",
		413: "Payload Too Large",
		414: "URI Too Long",
		415: "Unsupported Media Type",
		416: "Range Not Satisfiable",
		417: "Expectation Failed",
		418: "Im a teapot",
		421: "Misdirected Request",
		422: "Unprocessable Content",
		423: "Locked",
		424: "Failed Dependency",
		425: "Too Early",
		426: "Upgrade Required",
		428: "Precondition Required",
		429: "Too Many Requests",
		431: "Request Header Fields Too Large",
		451: "Unavailable For Legal Reasons",
		500: "Internal Server Error",
		501: "Not Implemented",
		502: "Bad Gateway",
		503: "Service Unavailable",
		504: "Gateway Timeout",
		505: "HTTP Version Not Supported",
		506: "Variant Also Negotiates",
		507: "Insufficient Storage",
		508: "Loop Detected",
		510: "Not Extended",
		511: "Network Authentication Required",
		...options.errors
	};
	const error = errors[result.status];
	if (error) throw new ApiError(options, result, error);
	if (!result.ok) {
		const errorStatus = result.status ?? "unknown";
		const errorStatusText = result.statusText ?? "unknown";
		const errorBody = (() => {
			try {
				return JSON.stringify(result.body, null, 2);
			} catch (e) {
				return void 0;
			}
		})();
		throw new ApiError(options, result, `Generic Error: status: ${errorStatus}; status text: ${errorStatusText}; body: ${errorBody}`);
	}
};
/**
* Request method
* @param config The OpenAPI configuration object
* @param options The request options from the service
* @returns CancelablePromise<T>
* @throws ApiError
*/
const request = (config, options) => {
	return new CancelablePromise(async (resolve$1, reject, onCancel) => {
		try {
			const url = getUrl(config, options);
			const formData = getFormData(options);
			const body = getRequestBody(options);
			const headers = await getHeaders(config, options);
			if (!onCancel.isCancelled) {
				let response = await sendRequest(config, options, url, body, formData, headers, onCancel);
				for (const fn of config.interceptors.response._fns) response = await fn(response);
				const responseBody = await getResponseBody(response);
				const responseHeader = getResponseHeader(response, options.responseHeader);
				const result = {
					url,
					ok: response.ok,
					status: response.status,
					statusText: response.statusText,
					body: responseHeader ?? responseBody
				};
				catchErrorCodes(options, result);
				resolve$1(result.body);
			}
		} catch (error) {
			reject(error);
		}
	});
};

//#endregion
export { request };
