// windmill-client/dist/index.mjs
import { ApiError } from "./core/ApiError.mjs";
import { CancelError, CancelablePromise } from "./core/CancelablePromise.mjs";
import { OpenAPI } from "./core/OpenAPI.mjs";
import { AdminService, AgentWorkersService, AppService, AssetService, AuditService, CaptureService, ConcurrencyGroupsService, ConfigService, DocumentationService, DraftService, EmailTriggerService, FavoriteService, FlowConversationService, FlowService, FolderService, GcpTriggerService, GitSyncService, GranularAclService, GroupService, HealthService, HelpersService, HttpTriggerService, IndexSearchService, InputService, IntegrationService, JobService, KafkaTriggerService, McpOauthService, McpService, MetricsService, MqttTriggerService, NativeTriggerService, NatsTriggerService, NpmProxyService, OauthService, OidcService, OpenapiService, PostgresTriggerService, RawAppService, ResourceService, ScheduleService, ScriptService, ServiceLogsService, SettingService, SettingsService, SqsTriggerService, TeamsService, TokenService, TriggerService, UserService, VariableService, VolumeService, WebsocketTriggerService, WorkerService, WorkspaceDependenciesService, WorkspaceIntegrationService, WorkspaceService } from "./services.gen.mjs";
import { datatable, ducklake } from "./sqlUtils.mjs";
import { SHARED_FOLDER, StepSuspend, WorkflowCtx, _workflowCtx, appendToResultStream, base64ToUint8Array, commitKafkaOffsets, databaseUrlFromResource, denoS3LightClientSettings, getFlowUserState, getIdToken, getInternalState, getPresignedS3PublicUrl, getPresignedS3PublicUrls, getProgress, getResource, getResult, getResultMaybe, getResumeEndpoints, getResumeUrls, getRootJobId, getState, getStatePath, getVariable, getWorkspace, loadS3File, loadS3FileStream, parallel, parseS3Object, requestInteractiveSlackApproval, requestInteractiveTeamsApproval, resolveDefaultResource, runFlow, runFlowAsync, runScript, runScriptAsync, runScriptByHash, runScriptByHashAsync, runScriptByPath, runScriptByPathAsync, setClient, setFlowUserState, setInternalState, setProgress, setResource, setState, setVariable, setWorkflowCtx, signS3Object, signS3Objects, sleep, step, streamResult, task, taskFlow, taskScript, uint8ArrayToBase64, usernameToEmail, waitForApproval, waitJob, workflow, writeS3File } from "./client.mjs";

//#region src/index.ts
const wmill = {
	setClient,
	getVariable,
	setVariable,
	getResource,
	setResource,
	getResumeUrls,
	setState,
	setProgress,
	getProgress,
	getState,
	getIdToken,
	denoS3LightClientSettings,
	loadS3FileStream,
	loadS3File,
	writeS3File,
	signS3Objects,
	signS3Object,
	getPresignedS3PublicUrls,
	getPresignedS3PublicUrl,
	task,
	taskScript,
	taskFlow,
	workflow,
	step,
	sleep,
	parallel,
	waitForApproval,
	WorkflowCtx,
	_workflowCtx,
	setWorkflowCtx,
	StepSuspend,
	runScript,
	runScriptAsync,
	runScriptByPath,
	runScriptByHash,
	runScriptByPathAsync,
	runScriptByHashAsync,
	runFlow,
	runFlowAsync,
	waitJob,
	getRootJobId,
	setFlowUserState,
	getFlowUserState,
	usernameToEmail,
	requestInteractiveSlackApproval,
	requestInteractiveTeamsApproval,
	appendToResultStream,
	streamResult,
	datatable,
	ducklake,
	SHARED_FOLDER,
	getWorkspace,
	getStatePath,
	getInternalState,
	setInternalState,
	getResumeEndpoints,
	getResult,
	getResultMaybe,
	resolveDefaultResource,
	databaseUrlFromResource,
	base64ToUint8Array,
	uint8ArrayToBase64,
	parseS3Object,
	commitKafkaOffsets,
	AdminService,
	AuditService,
	FlowService,
	GranularAclService,
	GroupService,
	JobService,
	ResourceService,
	VariableService,
	ScriptService,
	ScheduleService,
	SettingsService,
	UserService,
	WorkspaceService,
	TeamsService
};
var src_default = wmill;

//#endregion
export { AdminService, AgentWorkersService, ApiError, AppService, AssetService, AuditService, CancelError, CancelablePromise, CaptureService, ConcurrencyGroupsService, ConfigService, DocumentationService, DraftService, EmailTriggerService, FavoriteService, FlowConversationService, FlowService, FolderService, GcpTriggerService, GitSyncService, GranularAclService, GroupService, HealthService, HelpersService, HttpTriggerService, IndexSearchService, InputService, IntegrationService, JobService, KafkaTriggerService, McpOauthService, McpService, MetricsService, MqttTriggerService, NativeTriggerService, NatsTriggerService, NpmProxyService, OauthService, OidcService, OpenAPI, OpenapiService, PostgresTriggerService, RawAppService, ResourceService, ScheduleService, ScriptService, ServiceLogsService, SettingService, SettingsService, SqsTriggerService, StepSuspend, TeamsService, TokenService, TriggerService, UserService, VariableService, VolumeService, WebsocketTriggerService, WorkerService, WorkflowCtx, WorkspaceDependenciesService, WorkspaceIntegrationService, WorkspaceService, _workflowCtx, appendToResultStream, commitKafkaOffsets, datatable, src_default as default, denoS3LightClientSettings, ducklake, getFlowUserState, getIdToken, getPresignedS3PublicUrl, getPresignedS3PublicUrls, getProgress, getResource, getResumeUrls, getRootJobId, getState, getVariable, loadS3File, loadS3FileStream, parallel, requestInteractiveSlackApproval, requestInteractiveTeamsApproval, runFlow, runFlowAsync, runScript, runScriptAsync, runScriptByHash, runScriptByHashAsync, runScriptByPath, runScriptByPathAsync, setClient, setFlowUserState, setProgress, setResource, setState, setVariable, setWorkflowCtx, signS3Object, signS3Objects, sleep, step, streamResult, task, taskFlow, taskScript, usernameToEmail, waitForApproval, waitJob, workflow, writeS3File };
