// windmill-client/dist/sqlUtils.mjs
import { JobService } from "./services.gen.mjs";
import { getWorkspace } from "./client.mjs";

//#region src/sqlUtils.ts
/**
* Create a SQL template function for PostgreSQL/datatable queries
* @param name - Database/datatable name (default: "main")
* @returns SQL template function for building parameterized queries
* @example
* let sql = wmill.datatable()
* let name = 'Robin'
* let age = 21
* await sql`
*   SELECT * FROM friends
*     WHERE name = ${name} AND age = ${age}::int
* `.fetch()
*/
function datatable(name = "main") {
	return sqlProviderImpl("datatable", parseName(name));
}
/**
* Create a SQL template function for DuckDB/ducklake queries
* @param name - DuckDB database name (default: "main")
* @returns SQL template function for building parameterized queries
* @example
* let sql = wmill.ducklake()
* let name = 'Robin'
* let age = 21
* await sql`
*   SELECT * FROM friends
*     WHERE name = ${name} AND age = ${age}
* `.fetch()
*/
function ducklake(name = "main") {
	return sqlProviderImpl("ducklake", { name });
}
function sqlProviderImpl(provider, { name, schema }) {
	let sqlFn = (strings, ...values) => {
		let formatArgDecl = {
			datatable: (i) => `-- $${i + 1} arg${i + 1}`,
			ducklake: (i) => {
				let argType = parseTypeAnnotation(strings[i], strings[i + 1]) || inferSqlType(values[i]);
				return `-- $arg${i + 1} (${argType})`;
			}
		}[provider];
		let formatArgUsage = {
			datatable: (i) => {
				const parsedType = parseTypeAnnotation(strings[i], strings[i + 1]);
				if (parsedType !== void 0) return `$${i + 1}`;
				let argType = inferSqlType(values[i]);
				return `$${i + 1}::${argType}`;
			},
			ducklake: (i) => `$arg${i + 1}`
		}[provider];
		let content = values.map((_, i) => formatArgDecl(i)).join("\n") + "\n";
		if (provider === "ducklake") content += `ATTACH 'ducklake://${name}' AS dl;USE dl;\n`;
		if (schema && provider === "datatable") content += `SET search_path TO "${schema}";\n`;
		let contentBody = "";
		for (let i = 0; i < strings.length; i++) {
			contentBody += strings[i];
			if (i !== strings.length - 1) contentBody += formatArgUsage(i);
		}
		content += contentBody;
		const args = {
			...Object.fromEntries(values.map((v, i) => [`arg${i + 1}`, v])),
			...provider === "datatable" ? { database: `datatable://${name}` } : {}
		};
		const language = {
			datatable: "postgresql",
			ducklake: "duckdb"
		}[provider];
		async function fetch({ resultCollection } = {}) {
			if (resultCollection) content = `-- result_collection=${resultCollection}\n${content}`;
			try {
				let result = await JobService.runScriptPreviewInline({
					workspace: getWorkspace(),
					requestBody: {
						args,
						content,
						language
					}
				});
				return result;
			} catch (e) {
				let err = e;
				if (e && typeof e.body == "string" && e.statusText == "Internal Server Error") {
					let body = e.body;
					if (body.startsWith("Internal:")) body = body.slice(9).trim();
					if (body.startsWith("Error:")) body = body.slice(6).trim();
					if (body.startsWith("datatable")) body = body.slice(9).trim();
					err = Error(`${provider} ${body}`);
					err.query = contentBody;
					err.request = e.request;
				}
				throw err;
			}
		}
		return {
			content,
			args,
			fetch,
			fetchOne: (params) => fetch({
				...params,
				resultCollection: "last_statement_first_row"
			}),
			fetchOneScalar: (params) => fetch({
				...params,
				resultCollection: "last_statement_first_row_scalar"
			}),
			execute: (params) => fetch(params)
		};
	};
	if (provider === "datatable") sqlFn.query = (sqlString, ...params) => {
		let arr = Object.assign([sqlString], { raw: [sqlString] });
		return sqlFn(arr, ...params);
	};
	return sqlFn;
}
function inferSqlType(value) {
	if (typeof value === "number" || typeof value === "bigint") {
		if (Number.isInteger(value)) return "BIGINT";
		return "DOUBLE PRECISION";
	} else if (value === null || value === void 0) return "TEXT";
	else if (typeof value === "string") return "TEXT";
	else if (typeof value === "object") return "JSON";
	else if (typeof value === "boolean") return "BOOLEAN";
	else return "TEXT";
}
function parseTypeAnnotation(prevTemplateString, nextTemplateString) {
	if (!nextTemplateString) return;
	nextTemplateString = nextTemplateString.trimStart();
	if (nextTemplateString.startsWith("::")) return nextTemplateString.substring(2).trimStart().split(/\s+/)[0];
	prevTemplateString = prevTemplateString?.trimEnd();
	if (prevTemplateString?.endsWith("(") && prevTemplateString.substring(0, prevTemplateString.length - 1).trim().toUpperCase().endsWith("CAST") && nextTemplateString.toUpperCase().startsWith("AS ")) return nextTemplateString.substring(2).trimStart().split(/\s+/)[0];
}
function parseName(name) {
	if (!name) return { name: "main" };
	let [assetName, schemaName] = name.split(":");
	if (schemaName) return {
		name: assetName || "main",
		schema: schemaName
	};
	else return { name };
}

//#endregion
export { datatable, ducklake };
